src/common/asio_helpers.h

```diff
@@ -0,0 +1,141 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2017 Red Hat, Inc. <contact@redhat.com>
+ *
+ * Author: Adam C. Emerson <aemerson@redhat.com>
+ *
+ * Based on a prototype by Casey Bodley <cbodley@redhat.com>
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License version
+ * 2.1, as published by the Free Software Foundation.  See file
+ * COPYING.
+ */
+
+#ifndef CEPH_COMMON_ASIO_HELPERS_H
+#define CEPH_COMMON_ASIO_HELPERS_H
+
+#include <new>
+
+#define ASIO_STANDALONE 1
+#include <asio.hpp>
+
+#include <common/backport14.h>
+
+namespace ceph {
+
+namespace {
+/// map negative return values to error codes
+inline std::error_code to_error_code(int ret)
+{
+  if (ret < 0)
+    return {-ret, std::generic_category()};
+  else
+    return {};
+}
+}
+
+template <typename...>
+class invoker;
+
+template <typename R, typename... Args>
+class invoker<R(Args...)> {
+  class base {
+  public:
+    base() = default;
+    virtual R invoke(Args ...args) = 0;
+    virtual void destroy() = 0;
+    virtual void deallocate() = 0;
+    virtual base* move(void* p) = 0;
+
+  protected:
+    virtual ~base() = default;
+  };
+
+  template<typename Handler>
+  class derived : public base {
+    Handler h;
+  public:
+    derived(const Handler& h) : h(h) {}
+    derived(Handler&& h) : h(std::move(h)) {}
+
+    derived& operator =(const derived& h) = delete;
+    derived& operator =(derived&& h) = delete;
+
+    R invoke(Args ...args) override {
+      asio_handler_invoke_helpers::invoke(
+	std::bind(h, std::forward<Args>(args)...), h);
+    }
+    void destroy() override {
+      this->~derived();
+    }
+    void deallocate() override {
+      delete this;
+    }
+    base* move(void* p) override {
+      new(p) derived(std::move(h));
+      this->~derived();
+      return static_cast<base*>(p);
+    }
+  protected:
+    virtual ~derived() = default;
+  };
+
+  mutable ceph::aligned_storage_t<sizeof(void*) * 4> buf;
+  mutable base* specific;
+
+  bool sfo() const {
+    return static_cast<void*>(specific) == static_cast<void*>(&buf);
+  }
+ public:
+  invoker() = delete;
+  template<typename Handler>
+  invoker(Handler&& handler)
+    : specific(static_cast<base*>(
+		 sizeof(Handler) > sizeof(buf) ||
+		 !std::is_nothrow_move_constructible<Handler>{} ?
+		new derived<Handler>(std::forward<Handler>(handler)) :
```

Comment:
asio provides `boost_asio_handler_alloc_helpers` for per-handler allocations that can support some neat optimizations. taking advantage of those can be future work (and will probably live in layers above librados), but i think it's worth building in support for that now

it does impose a big restriction though, which is that the memory allocated with these helpers must be released before the handler is invoked. that would, for example, allow a handler to reschedule an operation and make use of the same memory. in my prototype, this was accomplished with [make_notifier()](https://github.com/cbodley/ceph/commit/a05e541d7456acba74427edca2e28a2dba8bf4e4#diff-ecddf235281a6fd977ee2041965364faR137) and [struct delete_and_invoke_notifier](https://github.com/cbodley/ceph/commit/a05e541d7456acba74427edca2e28a2dba8bf4e4#diff-ecddf235281a6fd977ee2041965364faR75)

the boost docs include a [custom_alloc_handler example](http://www.boost.org/doc/libs/1_64_0/doc/html/boost_asio/example/cpp11/allocation/server.cpp) that shows how a caller with some knowledge of the allocations required down the stack can preallocate memory to serve those allocations